// mongodb-memory-server-core/lib/util/resolveConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.envToBool = exports.envName = exports.resolveConfig = exports.findPackageJson = exports.setDefaultValue = exports.defaultValues = exports.ENV_CONFIG_PREFIX = exports.ResolveConfigVariables = void 0;
const tslib_1 = require("tslib");
const camelcase_1 = tslib_1.__importDefault(require("camelcase"));
const new_find_package_json_1 = require("new-find-package-json");
const debug_1 = tslib_1.__importDefault(require("debug"));
const path = tslib_1.__importStar(require("path"));
const fs_1 = require("fs");
const log = debug_1.default('MongoMS:ResolveConfig');
var ResolveConfigVariables;
(function (ResolveConfigVariables) {
    ResolveConfigVariables["DOWNLOAD_DIR"] = "DOWNLOAD_DIR";
    ResolveConfigVariables["PLATFORM"] = "PLATFORM";
    ResolveConfigVariables["ARCH"] = "ARCH";
    ResolveConfigVariables["VERSION"] = "VERSION";
    ResolveConfigVariables["DEBUG"] = "DEBUG";
    ResolveConfigVariables["DOWNLOAD_MIRROR"] = "DOWNLOAD_MIRROR";
    ResolveConfigVariables["DOWNLOAD_URL"] = "DOWNLOAD_URL";
    ResolveConfigVariables["PREFER_GLOBAL_PATH"] = "PREFER_GLOBAL_PATH";
    ResolveConfigVariables["DISABLE_POSTINSTALL"] = "DISABLE_POSTINSTALL";
    ResolveConfigVariables["SYSTEM_BINARY"] = "SYSTEM_BINARY";
    ResolveConfigVariables["MD5_CHECK"] = "MD5_CHECK";
    ResolveConfigVariables["ARCHIVE_NAME"] = "ARCHIVE_NAME";
    ResolveConfigVariables["RUNTIME_DOWNLOAD"] = "RUNTIME_DOWNLOAD";
    ResolveConfigVariables["USE_HTTP"] = "USE_HTTP";
})(ResolveConfigVariables = exports.ResolveConfigVariables || (exports.ResolveConfigVariables = {}));
exports.ENV_CONFIG_PREFIX = 'MONGOMS_';
exports.defaultValues = new Map([
    // apply app-default values here
    [ResolveConfigVariables.VERSION, '4.0.25'],
    [ResolveConfigVariables.PREFER_GLOBAL_PATH, 'true'],
    [ResolveConfigVariables.RUNTIME_DOWNLOAD, 'true'],
    [ResolveConfigVariables.USE_HTTP, 'false'],
]);
/**
 * Set an Default value for an specific key
 * Mostly only used internally (for the "global-x.x" packages)
 * @param key The Key the default value should be assigned to
 * @param value The Value what the default should be
 */
function setDefaultValue(key, value) {
    exports.defaultValues.set(key, value);
}
exports.setDefaultValue = setDefaultValue;
let packageJsonConfig = {};
/**
 * Find the nearest package.json (that has an non-empty config field) for the provided directory
 * @param directory Set an custom directory to search the config in (default: process.cwd())
 */
function findPackageJson(directory) {
    var _a, _b, _c;
    let filename;
    for (const filename of new_find_package_json_1.findSync(directory || process.cwd())) {
        log(`findPackageJson: Found package.json at "${filename}"`);
        const readout = JSON.parse(fs_1.readFileSync(filename).toString());
        if (Object.keys((_b = (_a = readout === null || readout === void 0 ? void 0 : readout.config) === null || _a === void 0 ? void 0 : _a.mongodbMemoryServer) !== null && _b !== void 0 ? _b : {}).length > 0) {
            log(`findPackageJson: Found package with non-empty config field at "${filename}"`);
            // the optional chaining is needed, because typescript wont accept an "isNullOrUndefined" in the if with "&& Object.keys"
            packageJsonConfig = (_c = readout === null || readout === void 0 ? void 0 : readout.config) === null || _c === void 0 ? void 0 : _c.mongodbMemoryServer;
            break;
        }
    }
    // block for all file-path resolving
    if (filename) {
        // These are so that "camelCase" doesnt get executed much & de-duplicate code
        // "cc*" means "camelcase"
        const ccDownloadDir = camelcase_1.default(ResolveConfigVariables.DOWNLOAD_DIR);
        const ccSystemBinary = camelcase_1.default(ResolveConfigVariables.SYSTEM_BINARY);
        if (ccDownloadDir in packageJsonConfig) {
            packageJsonConfig[ccDownloadDir] = path.resolve(path.dirname(filename), packageJsonConfig[ccDownloadDir]);
        }
        if (ccSystemBinary in packageJsonConfig) {
            packageJsonConfig[ccSystemBinary] = path.resolve(path.dirname(filename), packageJsonConfig[ccSystemBinary]);
        }
    }
    return packageJsonConfig;
}
exports.findPackageJson = findPackageJson;
/**
 * Resolve "variableName" value (process.env | packagejson | default | undefined)
 * @param variableName The variable to search an value for
 */
function resolveConfig(variableName) {
    var _a, _b, _c;
    return (_c = ((_b = (_a = process.env[envName(variableName)]) !== null && _a !== void 0 ? _a : packageJsonConfig[camelcase_1.default(variableName)]) !== null && _b !== void 0 ? _b : exports.defaultValues.get(variableName))) === null || _c === void 0 ? void 0 : _c.toString();
}
exports.resolveConfig = resolveConfig;
exports.default = resolveConfig;
/**
 * Helper Function to add the prefix for "process.env[]"
 */
function envName(variableName) {
    return `${exports.ENV_CONFIG_PREFIX}${variableName}`;
}
exports.envName = envName;
/**
 * Convert "1, on, yes, true" to true (otherwise false)
 * @param env The String / Environment Variable to check
 */
function envToBool(env = '') {
    if (typeof env !== 'string') {
        log('envToBool: input was not an string!');
        return false;
    }
    return ['1', 'on', 'yes', 'true'].indexOf(env.toLowerCase()) !== -1;
}
exports.envToBool = envToBool;
// enable debug if "MONGOMS_DEBUG" is true
if (envToBool(resolveConfig(ResolveConfigVariables.DEBUG))) {
    debug_1.default.enable('MongoMS:*');
}
// run this always after debug is enabled
findPackageJson();
